README.md
# Perforce Cookbook

[![Cookbook Version](http://img.shields.io/cookbook/v/perforce.svg?style=flat-square)][supermarket]
[![linux](http://img.shields.io/travis/dhoer/chef-perforce/master.svg?label=linux&style=flat-square)][linux]
[![osx](http://img.shields.io/travis/dhoer/chef-perforce/macosx.svg?label=macosx&style=flat-square)][osx]
[![win](https://img.shields.io/appveyor/ci/dhoer/chef-perforce/master.svg?label=windows&style=flat-square)][win]

[supermarket]: https://supermarket.chef.io/cookbooks/perforce
[linux]: https://travis-ci.org/dhoer/chef-perforce/branches
[osx]: https://travis-ci.org/dhoer/chef-perforce/branches
[win]: https://ci.appveyor.com/project/dhoer/chef-perforce

Installs Perforce Command-Line Client (p4).

## Requirements

- Chef 12.6+

### Platforms

- CentOS/RedHat/Fedora 
- Debian/Ubuntu
- Mac OS X
- Windows

### Cookbooks

- apt
- yum 

## Usage

Include recipe in cookbook or run list to install Perforce CLI.

## Getting Help

- Ask specific questions on [Stack Overflow](http://stackoverflow.com/questions/tagged/chef+perforce).
- Report bugs and discuss potential features in [Github issues](https://github.com/dhoer/chef-perforce/issues).

## Contributing

Please refer to [CONTRIBUTING](https://github.com/dhoer/chef-perforce/blob/master/CONTRIBUTING.md).

## License

MIT - see the accompanying [LICENSE](https://github.com/dhoer/chef-perforce/blob/master/LICENSE.md) file for details.


CHANGELOG.md
# CHANGELOG

## 2.1.0 2016-11-14

- Support new cli package name

## 2.0.0 2016-10-04

- Drop support for Chef 11

## 1.4.1 2016-09-16

- Debian 8 should use trusty package
- Add support for Ubuntu 13 and 15 versions

## 1.4.0 2016-09-16

- Use p4 r16.1 release
- Add Mac OS X support

## 1.3.0 2016-07-07

- Add Fedora 19+ support
- Add Ubuntu 16.04 support

## 1.2.4 2016-01-22

- Relax pessimistic locking on apt and yum dependencies 

## 1.2.3 2016-01-21

- Fix #2 Incompatible windows version

## 1.2.2 2015-12-28

- Change windows_package name to match registry DisplayName to prevent re-installs

## 1.2.1 2015-12-16

- Fix Windows unattended install

## 1.2.0 2015-11-27

- Add Windows support

## 1.1.0 2015-11-27

- Add Debian support

## 1.0.0 2015-10-27

- Initial release


spec/unit/spec_helper.rb
# Encoding: utf-8
require 'chefspec'
require 'chefspec/berkshelf'
require 'chefspec/cacher'

ChefSpec::Coverage.start!

::LOG_LEVEL = :warn
::CENTOS_OPTS = {
  platform: 'centos',
  version: '6.6',
  log_level: ::LOG_LEVEL
}.freeze
::UBUNTU_OPTS = {
  platform: 'ubuntu',
  version: '14.04',
  log_level: ::LOG_LEVEL
}.freeze
::WINDOWS_OPTS = {
  platform: 'windows',
  version: '2012R2',
  log_level: ::LOG_LEVEL
}.freeze
::CHEFSPEC_OPTS = {
  log_level: ::LOG_LEVEL
}.freeze


spec/unit/default_spec.rb
# Encoding: utf-8

require_relative 'spec_helper'

describe 'perforce::default' do
  describe 'rhel' do
    let(:chef_run) do
      ChefSpec::ServerRunner.new(platform: 'centos',
                                 version: '6.6',
                                 log_level: ::LOG_LEVEL).converge(described_recipe)
    end

    it 'adds p4 repo' do
      expect(chef_run).to create_yum_repository('Perforce')
    end

    it 'installs p4 cli package' do
      expect(chef_run).to install_package('helix-cli')
    end
  end

  describe 'debian' do
    let(:chef_run) do
      ChefSpec::ServerRunner.new(platform: 'ubuntu',
                                 version: '14.04',
                                 log_level: ::LOG_LEVEL).converge(described_recipe)
    end

    it 'adds p4 repo' do
      expect(chef_run).to add_apt_repository('Perforce')
    end

    it 'installs p4 command-line client package' do
      expect(chef_run).to install_package('helix-cli')
    end
  end

  describe 'windows' do
    let(:chef_run) do
      ChefSpec::ServerRunner.new(platform: 'windows',
                                 version: '2012R2',
                                 log_level: ::LOG_LEVEL).converge(described_recipe)
    end

    it 'installs p4 command-line client package' do
      expect(chef_run).to install_windows_package('Helix P4 Command-Line Client')
    end
  end

  describe 'mac_os_x' do
    let(:chef_run) do
      ChefSpec::ServerRunner.new(platform: 'mac_os_x',
                                 version: '10.11.1',
                                 log_level: ::LOG_LEVEL).converge(described_recipe)
    end

    it 'creates perforce directory' do
      expect(chef_run).to create_directory('/opt/perforce').with(
        user: 'root',
        group: 'wheel'
      )
    end
    it 'installs p4 command-line client package' do
      expect(chef_run).to create_remote_file('/opt/perforce/p4')
    end
    it 'creates PATH extension directory' do
      expect(chef_run).to create_directory('/etc/paths.d').with(
        user: 'root',
        group: 'wheel'
      )
    end
    it 'installs PATH extension file' do
      expect(chef_run).to create_file('/etc/paths.d/perforce').with(
        user: 'root',
        group: 'wheel'
      )
    end
  end
end


test/integration/default/serverspec/default_spec.rb
# Encoding: utf-8
require 'serverspec'

if (/cygwin|mswin|mingw|bccwin|wince|emx/ =~ RUBY_PLATFORM).nil?
  set :backend, :exec
else
  set :backend, :cmd
  set :os, family: 'windows'
end

case os[:family]
when 'windows'
  describe command('cmd /c "C:\Program Files\Perforce\p4.exe"') do
    its(:stderr) { should match(/Perforce client/) }
  end
# TODO: https://github.com/test-kitchen/test-kitchen/issues/1120
when 'darwin'
  describe command('/opt/perforce/p4 2>&1') do
    its(:stdout) { should match(/Perforce client/) }
  end
else # linux
  describe command('p4 2>&1') do
    its(:stdout) { should match(/Perforce client/) }
  end
end


recipes/default.rb
major_version = node['platform_version'].split('.')[0] # returns major version, e.g., version 6.4 would return 6

# https://www.perforce.com/perforce-packages
case node['platform_family']
when 'rhel', 'fedora'
  major_version = major_version.to_i < 19 ? raise('Only Fedora 19+ supported!') : '7' if node['platform'] == 'fedora'

  yum_repository 'Perforce' do
    description 'Perforce Repo'
    baseurl "http://package.perforce.com/yum/rhel/#{major_version}/x86_64"
    gpgkey 'https://package.perforce.com/perforce.pubkey'
    action :create
  end

  package 'perforce-cli'
when 'debian'
  dist = case major_version # http://askubuntu.com/a/445496
         when '12', '13', '7'
           'precise'
         when '14', '15', '8'
           'trusty'
         else # Assume Ubuntu 16+ or Debian 9+
           'xenial'
         end

  apt_repository 'Perforce' do
    uri 'http://package.perforce.com/apt/ubuntu'
    distribution dist
    components ['release']
    key 'https://package.perforce.com/perforce.pubkey'
  end

  package 'perforce-cli'
when 'windows'
  bit = node['kernel']['machine'] == 'x86_64' ? 'x64' : 'x86'

  windows_package 'Helix P4 Command-Line Client' do
    source "#{node['perforce']['download_url']}/#{node['perforce']['release']}/bin.nt#{bit}/helix-p4-#{bit}.exe"
    options '/v"/qn"'
    installer_type :custom
  end
when 'mac_os_x'
  bit = node['kernel']['machine'] == 'x86_64' ? 'x86_64' : 'x86'

  # We put p4 in /opt/perforce and not in /usr/local/bin as that space is locked down by OSX 10.11 and above
  directory '/opt/perforce' do
    owner 'root'
    group 'wheel'
    mode '00755'
    action :create
    recursive true
  end

  remote_file '/opt/perforce/p4' do
    source "#{node['perforce']['download_url']}/#{node['perforce']['release']}/bin.macosx105#{bit}/p4"
    owner 'root'
    group 'wheel'
    mode '00755'
    action :create
  end

  # The below path exists on El Capitan but no longer exists in Sierra but continues to behave in the same way
  directory '/etc/paths.d' do
    owner 'root'
    group 'wheel'
    mode '00755'
    action :create
    recursive true
  end

  file '/etc/paths.d/perforce' do
    content '/opt/perforce'
    owner 'root'
    group 'wheel'
    mode '00755'
    action :create
  end
else
  raise("Platform #{node['platform']} not supported!")
end
